1-E-Commerce/index.html

```diff
@@ -0,0 +1,258 @@
+            const item = products.find(item => item.id == itemId)
+
+            item['favorite'] = true
+
+            if (item) {
+                currentFavs.push(item)
+                localStorage.setItem('favorites', JSON.stringify(currentFavs))
+            }
+        }
+
+        const removeItemFromFavs = function (itemId) {
+            const currentFavs = JSON.parse(localStorage.getItem('favorites')) || []
+            const index = currentFavs.map(item => item.id).indexOf(itemId)
+            const item = products.find(item => item.id === itemId)
+
+            delete item['favorite']
+
+            if (index > -1) {
+                currentFavs.splice(index, 1)
+                localStorage.setItem('favorites', JSON.stringify(currentFavs))
+            }
+        }
+
+        fetch('https://dummyjson.com/products')
+            .then(res => res.json())
+            .then(data => {
+                products.push(...data.products);
+
+                const displayedOnLoad = products.slice(0, 6)
+
+                populateWithItems(displayedOnLoad)
+            })
+            .catch(err => console.log(err))
+
+        document.querySelector('#load-more')
+            .addEventListener('click', (e) => {
+            e.preventDefault()
+
+            clearItems()
+            augmentLimit()
+            const { limit } = localStorage
+
+            const newItems = products.slice(0, Number(limit))
+            populateWithItems(newItems)
+        })
+
+        document.querySelector('#searchbar')
+            .addEventListener('keyup', (e) => {
+                clearItems()
+
+                const inputValue = e.target.value
+
+                if (! inputValue) {
+                    localStorage.setItem('limit', '6')
+                    return populateWithItems(products.slice(0, 6))
+                }
+
+                const regex = new RegExp(inputValue, 'gi')
+
+                const filteredProducts = products.filter(item =>{
+                    return item.title.match(regex) || item.description.match(regex)
+                })
+
+                if (filteredProducts.length) return populateWithItems(filteredProducts)
+
+                document.querySelector('.products').innerHTML = `<h2>No Results Found</h2>`
+            })
+
+        document.querySelector('.products')
+            .addEventListener('click', (e) => {
+                const clickedItemIsFav = e.target.classList.contains('fav')
+                const clickedItemIsSVGPath = e.target.nodeName === 'path'
+
+                const productId = Number(e.target.dataset.product_id)
+
+                if (clickedItemIsFav) {
+                    e.target.classList.add('fav-red')
+                    return addItemToFavs(productId)
+                }
+                if (clickedItemIsSVGPath) {
+                    e.target.parentNode.classList.remove('fav-red')
+                    return removeItemFromFavs(productId)
+                }
+            })
```

Comment:
We should add the event listener to the actual button instead of the products div. Maybe in this case you can use something like `document.querySelectorAll('.fav')`, and then iterate over them adding the event listener.

Reply:
I think this might run into a problem. Every time I call `clearItems` and `populateWithItems` I'm changing the contents of `div.products`, so even if I've looped over all `.fav` icons, the event listeners won't register clicks on newly created cards. Another possible solution is to add an event listener on each created card node (probably in the function `populateWithItems`). What do you think is best?

PD: A subtle detail is that, when the heart icon is filled (colored red), clicking the heart icon will (almost always) be registered as a click on the `<path>` and not on the `<svg>` tag.

Reply:
I think I would declare a function `onFavoriteClick()` which will be the handler, and set to this div:

`<div class="card-fav-icon">` like this: `<button class="card-fav-icon" onClick="onFavoriteClick()">`

Note that I changed it to a `button`, since this is the only HTML element that we should add `onClick` to.

Then, instead of detecting whether the favorite is selected or not based on the classes or if it is a `path` node, we should keep this state ourselves. I would declare an array with all the favorite products, and then when `onFavoriteClick` is called, you can add/remove the product to this array.

Reply:
Or if you already have the localStorage array, we can use that one.